espnet/nets/pytorch_backend/transducer/decoder_interface.py

```diff
@@ -0,0 +1,109 @@
+"""Transducer decoder interface module."""
+
+from typing import Dict
+from typing import List
+from typing import Tuple
+from typing import Union
+
+import torch
+
+
+class TransducerDecoderInterface:
+    """Decoder interface for transducer models."""
+
+    def zero_state(
+        self, init_tensor: torch.Tensor = None
+    ) -> Union[Tuple[List[torch.Tensor], List[torch.Tensor]], List[torch.Tensor]]:
+        """Initialize decoder states.
+
+        Args:
+            init_tensor: input features
+
+        Returns:
+            state: initial state
+
+        """
+        raise NotImplementedError("zero_state method is not implemented")
+
+    def forward_one_step(
+        self,
+        hyp: Dict[str, Union[float, List, torch.Tensor, None]],
+        init_tensor: torch.Tensor = None,
+    ) -> Union[
+        List[torch.Tensor], Tuple[List[torch.Tensor], List[torch.Tensor]], torch.Tensor
+    ]:
+        """Forward one step.
+
+        Args:
+            hyp: hypothese
+            init_tensor: initial tensor for attention computation
+
+        Returns:
+            tgt: decoder outputs
+            new_state: new decoder state
+            att_w: attention weights
+            lm_tokens: input token id for LM
+
+        """
+        raise NotImplementedError("forward_one_step method is not implemented")
+
+    def forward_batch_one_step(
+        self,
+        hyps: List[Dict[str, Union[float, List, torch.Tensor, None]]],
+        state: Union[List, Tuple[List[torch.Tensor], List[torch.Tensor]]],
+        att_w: torch.Tensor = None,
+        att_params: List[Union[int, torch.Tensor]] = None,
+    ) -> Union[List, Tuple[List[torch.Tensor], List[torch.Tensor]], torch.Tensor]:
+        """Forward batch one step.
+
+        Args:
+            hyps: batch of hypothesis
+            state: batch of decoder states
+            att_w : batch of attention weights
+            att_params : attention parameters
+
+        Returns:
+            tgt: decoder outputs
+            new_state: new batch of decoder states
+            att_w: new batch of attention weights
+            lm_tokens: input token ids for LM
```

Comment:
I wonder if these `state, att_w, att_params` can be merged into one `state` or `new_state`. It looks cleaner for decoder models without `att_w` or `att_params`, e.g., DecoderTT.

Reply:
I merged them into a tuple, is that okay like that?